x-pack/plugins/security/server/lib/authentication/providers/token.js

```diff
@@ -0,0 +1,215 @@
+   */
+  constructor(options) {
+    this._options = options;
+  }
+
+  /**
+   * Performs request authentication using Basic HTTP Authentication.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @param {Object} [state] Optional state object associated with the provider.
+   * @returns {Promise.<AuthenticationResult>}
+   */
+  async authenticate(request, state) {
+    this._options.log(['debug', 'security', 'token'], `Trying to authenticate user request to ${request.url.path}.`);
+
+    // first try from login payload
+    let authenticationResult = await this._authenticateViaLoginAttempt(request);
+
+    // if there isn't a payload, try basic auth
+    if (authenticationResult.notHandled()) {
+      authenticationResult = await this._authenticateViaHeader(request);
+    }
+
+    // if we still can't attempt auth, try authenticating via state (session token)
+    if (authenticationResult.notHandled()) {
+      authenticationResult = await this._authenticateViaState(request, state);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = this._authenticateViaRefreshToken(request, state);
+      }
+    }
+
+    // finally, if authentication still can not be handled for this
+    // request/state combination, redirect to the login page if appropriate
+    if (authenticationResult.notHandled() && canRedirectRequest(request)) {
+      const nextURL = encodeURIComponent(`${this._options.basePath}${request.url.path}`);
+      return AuthenticationResult.redirectTo(
+        `${this._options.basePath}/login?next=${nextURL}`
+      );
+    }
+
+    return authenticationResult;
+  }
+
+  /**
+   * Redirects user to the login page preserving query string parameters.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<DeauthenticationResult>}
+   */
+  async deauthenticate(request) {
+    // Query string may contain the path where logout has been called or
+    // logout reason that login page may need to know.
+    return DeauthenticationResult.redirectTo(
+      `${this._options.basePath}/login${request.url.search}`
+    );
+  }
+
+  /**
+   * Validates whether request contains `Basic ***` Authorization header and just passes it
+   * forward to Elasticsearch backend.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<AuthenticationResult>}
+   * @private
+   */
+  async _authenticateViaHeader(request) {
+    this._options.log(['debug', 'security', 'token'], 'Trying to authenticate via header.');
+
+    const authorization = request.headers.authorization;
+    if (!authorization) {
+      this._options.log(['debug', 'security', 'token'], 'Authorization header is not presented.');
+      return AuthenticationResult.notHandled();
+    }
+
+    const authenticationSchema = authorization.split(/\s+/)[0];
+    if (authenticationSchema.toLowerCase() !== 'basic') {
```

Comment:
__nit:__ `basic` ---> `bearer` or you want this provider to support both or just `basic` for some reason? If someone is using `Basic ...` header directly I think they are supposed to use `basic` auth provider. Using `Bearer` here would allow us to use _any_ ES realm that understands tokens (e.g custom OAuth realms + OAuth proxy in front of Kibana).

Reply:
I double checked the SAML provider logic and sure enough it uses `bearer`. I was under the impression that we had hardcoded true basic auth header support into each provider to make sure people could still do API calls programmatically while also supporting UI-based auth workflows, but clearly that's not the case.

Unfortunately, this gets more tricky with basic/token providers. The login form itself can only work with one of those two providers, so I worry about the subtly in ordering of those providers in the configuration accidentally exposing people to issues they didn't realize. Like when you switch between token and basic auth for the login form, they will appear to work in exactly the same way, and you'd need to examine the contents of the encrypted cookie (not possible for end users) to even know which provider was handling login.

I wonder if it makes sense to introduce an explicit login handling configuration so you can configure things like basic and token providers at the same time while simultaneously ensuring that the login form only works with one specific provider.

Thoughts?

Reply:
The order of auth providers in `[basic, token]` or `[token, basic]` only influences login page, for programmatic API it won't matter (`basic` intentionally fails if it notices **non-empty** `Authorization` header that it doesn't understand to allow next realm to handle that, `token` should do the same).

Reply:
Just a couple of thoughts:

The order of auth providers in `[basic, token]` or `[token, basic]` only influences login page, for programmatic API it won't matter (`basic` intentionally fails if it notices **non-empty** `Authorization` header that it doesn't understand to allow next provider to handle that, `token` should do the same). For `[saml, token]` the order will matter though. But as far as I understand the order of realms can be tricky for ES as well and administrators should know what they do when they configure the order of realms. 

Having said that, I agree, "the first auth provider always manages login" rule isn't obvious, we can introduce config option to make it explicit which auth provider should handle login and then we can "sign" `loginAttempt` with provider name/id (like we do for session) based on this config. Otherwise there may be an issue when provider can try to use `loginAttempt` to authenticate user just because its shape looks like what provider expects even though it was supposed to be used by another provider (idk, like ldap-based auth provider that expects username and password and basic that expects the same).

Reply:
I feel like I'm missing something. I thought this provider was going to essentially be a "basic auth provider using tokens for sessions internally". In what situations would the requests from the end-users browser be presenting "bearer" authorization headers?

Reply:
@kobelb That was how I was thinking about it originally, but I'm coming around to the idea that the token provider is not a special basic auth provider at all. So the idea would be that if you want to support both token auth and basic auth, you'd have to enable both providers. You could lock out basic auth entirely if you wanted.

With this in mind, we don't _have_ to support header-based auth at all with the token provider. The only real use case I can think of would be people building custom SSO solutions (ldap perhaps?) that leverage the token provider rather than the basic provider as they do today. Correct me if I'm wrong about that possibility, though.

Reply:
@epixa gotcha, I'm with you all now, thanks for catching me up. 

Allowing custom SSO using the header makes sense conceptually, I don't see any real harm in including it if it opens up this possibility. I honestly haven't played around with various intelligent reverse proxies that would make this type of workflow possible enough to know the feasibility, but it'd be a ton nicer than the `run-as` approach that we have now.